### ISSUE TITLE
Output range changes when converted to TFLite

### ISSUE THREAD

**Issue by KevSr** [tensorflow/tensorflow#46926]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installation (pip package or built from source): Pip package
- TensorFlow library (version, if pip package or github SHA, if built from source): TF 1.14

### 2. Code

Provide code to help us reproduce your issues using one of the following options:

#### Code A: Tensorflow to TFLite conversion

```
def save(self):
        
        graph = tf.get_default_graph()
        with graph.as_default():
        if self.quant == True:
           tf.contrib.quantize.create_eval_graph()

        out_def = graph.as_graph_def()
        with tf.gfile.GFile(self.out_graph, 'wb') as f:
            f.write(out_def.SerializeToString())

def freeze(self):
        checkpoint = os.path.join(self.checkpoint_dir, 'Model%d' % self.trial)
        ckpt= tf.train.get_checkpoint_state(checkpoint)
        ckpt_name = os.path.basename(ckpt.model_checkpoint_path)
        ckpt_path = os.path.join(checkpoint, ckpt_name)

        self.out_name = 'output'
        freeze_graph.freeze_graph(self.out_graph,'',True,ckpt_path, self.out_name, restore_op_name='save/restore_all', filename_tensor_name='save/Const:0' , initializer_nodes='', output_graph= self.fr_graph ,clear_devices=True)


def convert(self):
        print(self.fr_graph)
        converter= tf.lite.TFLiteConverter.from_frozen_graph(graph_def_file= self.fr_graph, input_arrays= ['input'],\
                                                  input_shapes= {'input': self.input_shape}, output_arrays=[self.out_name])
        
        if self.quant == True:
            in_array = converter.get_input_arrays()
            converter.quantized_input_stats = {in_array[0] : (0, 1)}  # mean, std_dev (input range is [-1, 1])

            converter.inference_type = tf.uint8

            tflite_model = converter.convert()
            print('Save the model.')

            with open('int_qat_%d.tflite'%self.trial, 'wb') as f:
                f.write(tflite_model)

```

#### Option B:  TFLite inference code

```
def run_tflite(modelpath, labelpath, datapath, trial, height, weight):

    interpreter = tf.lite.Interpreter(model_path=modelpath)
    interpreter.allocate_tensors()

    # Get input and output tensors.
    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()

    # Test the model on random input data.

    in_s = input_details[0]['shape']

    num = len(label_list)
    tf_img, label, num = load_and_slice_img(img_list,label_list, num, height, weight)
    output_data= []
    for i in range(num):
        input_= tf_img[i].reshape((1, height//2, weight//2, 3))

        interpreter.set_tensor(input_details[0]['index'], input_)
        interpreter.invoke()

        out_data = interpreter.get_tensor(output_details[0]['index'])

        output_data.append(out_data)
    output_data = np.array(output_data).reshape((num, height, weight, 3))

```

### 3. Failure after conversion
If the conversion is successful, but the generated model is wrong, then state what is wrong:

I am trying to quantize a SR model (EDSR), using Quantization Aware Training.

I am using same codes used in Jacob et al. (2017) to perform quantization (create_training_graph etc)

Fake Quantization (Stimulated) works rather well, but when I convert them using TFLiteconverter (from freeze_model), the result becomes very poor (PSNR below that of bicubic interpolation)

When I draw histogram with the output and HR image, the range of image has decreased about 20% (e.g 0-200 to 25-180). Histogram structure seems to be identical.

I tried to cross-check this using a simple model for MNIST, and found out following things
- Almost zero performance change when converted to TFLite file, both with QAT and Post Integer Quantization
- Output, before applying softmax function, changes dramatically between TFLite and Float (real float and fake/stimulated quantization one) models.

I used images in raw form, without normalizing to [0,1] to train the floating point models. When converting the float model, I used exact codes written in relevant document in tf==1.14 github. I also used same codes for inference in TFLite

I would like to reproduce the issue, but since it requires a lot of computation to train SR model, I did not upload the file initially. If required though, I will try to upload snippets of MNIST code (+ conversion code) that I have used to produce this error
**Update - I have uploaded some codes and pb/tflite file of the models I produced**

When I had a look into each nodes, the only difference between tflite and QAT(float) was a small change in min/max values for 'act_quant' stage, when activation function is not applied.

What might be the possible cause of this error (happens in both classification and SR model)? One easy thought is that the input image might have slightly changed after converting the format into float32 (from uint8) for inference in floating model.
Also, I think it might be the case that random bias has been added.

![KakaoTalk_20210204_233632390](https://user-images.githubusercontent.com/29692100/106907712-e8ba7100-6741-11eb-94b4-3ad0aff0ab5d.jpg)

Top Image- TFLite model after quantization (QAT)
Bottom Image - HR image

![KakaoTalk_20210204_233632390_01](https://user-images.githubusercontent.com/29692100/106907721-eb1ccb00-6741-11eb-8717-b1e5a1041913.jpg)

Histogram of above image for TFLite model after QAT

![KakaoTalk_20210204_233632390_02](https://user-images.githubusercontent.com/29692100/106907723-ebb56180-6741-11eb-864d-f2d46f27c66e.jpg)

Histogram of same image for floating point model

It is ok if there is no exact solution, but I would like to know possible cause of this phenomenon, so that I can try to debug it.

* I used non keras layers to create the model, and also manually recreated depth_to_space layer of EDSR, as it was not supported in the version of TFLite converter I used.

[Graphs.zip](https://github.com/tensorflow/tensorflow/files/5929670/Graphs.zip)
frozen graph (fake quant)  and corresponding tflite file (qat)

**Comment by KevSr** [comment]
It might be similar issue to this one, as result between vanilla and Fake Quant model is not very different in my case as well, and also, output of the model (although this is not an output from one Conv2d op) changes when converted to TFLite.

https://github.com/tensorflow/tensorflow/issues/38845

Just for the information, the Post Integer Quantization using TFLite version for TF 1.14 resulted with even worse result than that of QAT

**Comment by amahendrakar** [comment]
@KevSr,
TensorFlow 1.x is not actively supported. Could you please update TensorFlow to the latest stable version v2.4.1 and check if you are facing the same issue. Thanks!

**Comment by KevSr** [comment]
I will try with newer version of Tensorflow at later times, but I need to use v1.14 for this model.

Thank you for the help anyways
